### PR TITLE
Support monitor store size on S3

### PR DESCRIPTION
**Pull request by lidezhu**
### What problem does this PR solve?

Issue Number: ref #6827 

Problem Summary: The used store size on S3 is absent.

### What is changed and how it works?
Manage store size on S3:
1. Manage all current DTFile ids in `StableDiskDelegator` using a mapping like `external_id->pair<can_gc, file_size>`;
2. At restart, we put all segments' DTFiles using `StableDiskDelegator::addRemoteDTFileIfNotExists`;
3. When create new DTFile, we add DTFiles using `StableDiskDelegator::addRemoteDTFileWithGCDisabled` and `StableDiskDelegator::enableGCForRemoteDTFile`;
4. In the background, we will get all current DTFile ids using `StableDiskDelegator::getAllRemoteDTFilesForGC` and remove the ones which are not in the valid external id set of `PageDirectory` using `StableDiskDelegator::removeRemoteDTFile`;

image

Manage the store size for different keyspace id:
1. For store size on S3, it is just the total size of all DTFiles under the target keyspace id;
2. For local size, it is hard to distinguish data between different keyspace id, so we use a weighted mean to get an estimated value;

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Storages/PathPool.h [hunk 1]
```diff
@@ -494,10 +511,14 @@ class StoragePathPool
     String database;
     String table;
 
-    // This mutex mainly used to protect the `dt_file_path_map`
+    KeyspaceID keyspace_id = NullspaceID;
+
+    // This mutex mainly used to protect the `dt_file_path_map` and `s3_dt_file_size_map`.
```
Comment (JaySon-Huang) [inline dbms/src/Storages/PathPool.h:516]:
```suggestion
    // This mutex mainly used to protect the `dt_file_path_map` and `remote_dt_file_size_map`.
```

File: dbms/src/Storages/PathPool.h [hunk 2]
```diff
@@ -480,6 +495,8 @@ class StoragePathPool
     };
     using LatestPathInfos = std::vector<LatestPathInfo>;
 
+    using RemoteDMFileSizeMap = std::unordered_map<UInt64, std::pair<bool, UInt64>>;
```
Comment (JaySon-Huang) [inline dbms/src/Storages/PathPool.h:498]:
```suggestion
    // A map of DMFileID -> { isGCEnabled, FileSize }
    using RemoteDMFileSizeMap = std::unordered_map<UInt64, std::pair<bool, UInt64>>;
```

File: dbms/src/Storages/PathCapacityMetrics.cpp [hunk 3]
```diff
@@ -119,6 +121,52 @@ void PathCapacityMetrics::freeUsedSize(std::string_view file_path, size_t used_b
...
+    iter.first->second += used_bytes;
+}
+
+void PathCapacityMetrics::freeRemoteUsedSize(KeyspaceID keyspace_id, size_t used_bytes)
+{
+    std::unique_lock<std::mutex> lock(mutex);
+    auto iter = keyspace_id_to_used_bytes.find(keyspace_id);
+    RUNTIME_CHECK(iter != keyspace_id_to_used_bytes.end());
+    iter->second -= used_bytes;
+    RUNTIME_CHECK(iter->second >= 0);
```
Comment (JaySon-Huang) [inline dbms/src/Storages/PathCapacityMetrics.cpp:139]:
```suggestion
    RUNTIME_CHECK_MSG(iter->second >= 0, iter->second, keyspace_id, used_bytes);
```

File: dbms/src/Storages/PathCapacityMetrics.cpp [hunk 4]
```diff
@@ -119,6 +121,52 @@ void PathCapacityMetrics::freeUsedSize(std::string_view file_path, size_t used_b
...
+    std::unique_lock<std::mutex> lock(mutex);
+    auto iter = keyspace_id_to_used_bytes.emplace(keyspace_id, 0);
+    iter.first->second += used_bytes;
+}
+
+void PathCapacityMetrics::freeRemoteUsedSize(KeyspaceID keyspace_id, size_t used_bytes)
+{
+    std::unique_lock<std::mutex> lock(mutex);
+    auto iter = keyspace_id_to_used_bytes.find(keyspace_id);
+    RUNTIME_CHECK(iter != keyspace_id_to_used_bytes.end());
```
Comment (JaySon-Huang) [inline dbms/src/Storages/PathCapacityMetrics.cpp:137]:
```suggestion
    RUNTIME_CHECK(iter != keyspace_id_to_used_bytes.end(), keyspace_id);
```

File: dbms/src/Storages/PathCapacityMetrics.cpp [hunk 5]
```diff
@@ -187,33 +235,45 @@ FsStats PathCapacityMetrics::getFsStats(bool finalize_capacity)
...
+    const double avail_rate = 1.0 * total_stat.avail_size / total_stat.capacity_size;
+    // Default threshold "schedule.low-space-ratio" in PD is 0.8, log warning message if avail ratio is low.
+    if (avail_rate <= 0.2)
+        LOG_WARNING(
+            log,
+            "Available space is only {:.2f}% of capacity size. Avail size: {}, used size: {}, capacity size: {}",
+            avail_rate * 100.0,
+            formatReadableSizeWithBinarySuffix(total_stat.avail_size),
+            formatReadableSizeWithBinarySuffix(total_stat.used_size),
+            formatReadableSizeWithBinarySuffix(total_stat.capacity_size));
```
Comment (JaySon-Huang) [inline dbms/src/Storages/PathCapacityMetrics.cpp:275]: I think this warning logging is more useful for "local disk" available rate is low. Move it back to before line 241?

File: dbms/src/Storages/DeltaMerge/DeltaMergeStore_Ingest.cpp [hunk 6]
```diff
@@ -686,6 +694,18 @@ void DeltaMergeStore::ingestFiles(
     // deleted DTFiles to other segments.
     for (const auto & file : files)
         file->enableGC();
+
+    if (auto data_store = dm_context->db_context.getSharedContextDisagg()->remote_data_store; !data_store)
+    {
+        for (auto & file : files)
+            file->enableGC();
```
Comment (JaySon-Huang) [inline dbms/src/Storages/DeltaMerge/DeltaMergeStore_Ingest.cpp:701]: And we can remove line 695~696

File: dbms/src/Storages/DeltaMerge/DeltaMergeStore_InternalBg.cpp [hunk 7]
```diff
@@ -174,15 +175,96 @@ class LocalDMFileGcRemover final
...
+            return;
+
+        auto delegate = path_pool->getStableDiskDelegator();
+        for (const auto & [path, ids] : path_and_ids_vec)
+        {
+            for (auto id : ids)
+            {
+                if (!valid_ids.count(id))
+                {
+                    delegate.removeRemoteDTFile(id);
```
Comment (JaySon-Huang) [inline dbms/src/Storages/DeltaMerge/DeltaMergeStore_InternalBg.cpp:244]: maybe add some logging about removing the remote dtfile is helpful

File: dbms/src/Storages/PathCapacityMetrics.cpp (hunk 5 above)
Comment (lidezhu) [reply dbms/src/Storages/PathCapacityMetrics.cpp]: Fixed.

File: dbms/src/Storages/DeltaMerge/DeltaMergeStore_Ingest.cpp (hunk 6 above)
Comment (lidezhu) [reply dbms/src/Storages/DeltaMerge/DeltaMergeStore_Ingest.cpp:698]: Removed.

File: dbms/src/Storages/DeltaMerge/DeltaMergeStore_InternalBg.cpp (hunk 7 above)
Comment (lidezhu) [reply dbms/src/Storages/DeltaMerge/DeltaMergeStore_InternalBg.cpp:246]: Added.

**Review by JaySon-Huang — APPROVED** [review]
LGTM

**Pull request merged**